src/drivers/dw/dma.c

```diff
@@ -135,7 +135,7 @@ static struct dma_chan_data *dw_dma_channel_get(struct dma *dma,
 	trace_dwdma("dw_dma_channel_get(): dma %d request channel %d",
 		    dma->plat_data.id, req_chan);
 
-	spin_lock_irq(dma->lock, flags);
+	flags = spin_lock_irq(dma->lock);
 
```

Comment:
My eyes hurt looking at this. I do understand the need for replacing macros with functions, but I don't really like that we change the prototype.

Would it be possible to pass a pointer to flags as arguments, so that at least people used with locking API in kernel won't be confused about this.

So, instead of 
```flags = spin_lock_irq(lock)```
you will have
```
spin_lock_irq(lock, &flags);
```

Reply:
@dbaluta I don't like passing something that looks like passing by value, but in reality you pass it like reference and it gets modified (that's how old macros worked). How it in in kernel usually? It is like our old macros or with pointer? (pointer is ok for me, at least I can see that flags gonna be modified)

Reply:
@jajanusz in kernel is like our old approach with macros. What problems are we having with macros approach so that we need to change it, now?

Reply:
Problem - people sometimes pass wrong type of pointer and get no compilation warning and as I wrote previously you get nonexpected behaviour for your variables - you need to go one step deeper to understand code. That's what's wrong in this one case, I won't start religious war about why macros are even worse in other aspects  ;)

Reply:
@dbaluta And if it's really what's in kernel and if they keep using it, I can live with that, but if they are moving away from that and it's only legacy stuff then I'd prefer to handle it differently

Reply:
@jajanusz here is the implementation from Linux kernel:

https://elixir.bootlin.com/linux/latest/source/include/linux/spinlock.h#L247

Bottom line is that even they use macros they do some sort of type checking using this

https://elixir.bootlin.com/linux/latest/source/include/linux/typecheck.h#L9

Maybe we can do the same?

Reply:
@dbaluta Looks like an improvment while keeping our codebase to be more familiar for kernel developers, so it's good.
@bkokoszx Please just add checks then

Reply:
@dbaluta 
Thanks for examples.
@lgirdwood 
Is is ok for you to add typechecks macros?

Reply:
Lets not redesign, lets do it the Linux way with typecheck macros too. We can benefit from how this is done in Linux where it's probably been debated to death (and also validated with lots of architectures and compilers).

Reply:
@lgirdwood @paulstelian97 @bkokoszx so looking again at the spin_lock implementation in Linux its not that easy to follow but:
* type check for lock parameter is done because spin_lock_ macro eventually goes to an inline function call.
* type check for flags is done with the help of typecheck macro

So, @anyone feel free to jump with ideas of implementing this.

Reply:
@dbaluta what's our current delta with the Linux method ? My understanding was the latest updated aligned to Linux ?

Reply:
@lgirdwood @dbaluta 
ATM we do not have type checks for ```flags```. Type check for ```lock``` is done, because in our example ```spin_lock_``` macro also goes to an inline function call.